File: internal/service/sequence_svc.go
package service

import (
	"context"
	"errors"
	"fmt"

	"github.com/zajca/zfaktury/internal/domain"
	"github.com/zajca/zfaktury/internal/format"
	"github.com/zajca/zfaktury/internal/repository"
)

// SequenceService provides business logic for invoice sequence management.
type SequenceService struct {
	repo  repository.InvoiceSequenceRepo
	audit *AuditService
}

// NewSequenceService creates a new SequenceService.
func NewSequenceService(repo repository.InvoiceSequenceRepo, audit *AuditService) *SequenceService {
	return &SequenceService{repo: repo, audit: audit}
}

// Create validates uniqueness (prefix+year) within the company and persists a
// new invoice sequence.
func (s *SequenceService) Create(ctx context.Context, companyID int64, seq *domain.InvoiceSequence) error {
	if seq.Prefix == "" {
		return fmt.Errorf("prefix is required: %w", domain.ErrInvalidInput)
	}
	if seq.Year == 0 {
		return fmt.Errorf("year is required: %w", domain.ErrInvalidInput)
	}
	if seq.NextNumber <= 0 {
		seq.NextNumber = 1
	}
	if seq.FormatPattern == "" {
		seq.FormatPattern = "{prefix}{year}{number:04d}"
	}
	if err := format.ValidatePattern(seq.FormatPattern); err != nil {
		return fmt.Errorf("validating format pattern: %w", err)
	}

	// Check uniqueness of prefix+year within the company.
	existing, err := s.repo.GetByPrefixAndYear(ctx, companyID, seq.Prefix, seq.Year)
	if err == nil && existing != nil {
		return fmt.Errorf("sequence with prefix %q and year %d already exists: %w", seq.Prefix, seq.Year, domain.ErrDuplicateNumber)
	}
	// Only proceed if the error indicates not found.
	if err != nil && !errors.Is(err, domain.ErrNotFound) {
		return fmt.Errorf("checking sequence uniqueness: %w", err)
	}

	if err := s.repo.Create(ctx, companyID, seq); err != nil {
		return fmt.Errorf("creating sequence: %w", err)
	}
	if s.audit != nil {
		s.audit.Log(ctx, "sequence", seq.ID, "create", nil, seq)
	}
	return nil
}

// Update validates and updates an existing invoice sequence within the given company.
// Prevents lowering next_number below already-used numbers.
func (s *SequenceService) Update(ctx context.Context, companyID int64, seq *domain.InvoiceSequence) error {
	if seq.ID == 0 {
		return fmt.Errorf("sequence ID is required: %w", domain.ErrInvalidInput)
	}
	if seq.Prefix == "" {
		return fmt.Errorf("prefix is required: %w", domain.ErrInvalidInput)
	}
	if seq.Year == 0 {
		return fmt.Errorf("year is required: %w", domain.ErrInvalidInput)
	}
	if seq.NextNumber <= 0 {
		return fmt.Errorf("next number must be positive: %w", domain.ErrInvalidInput)
	}
	if seq.FormatPattern == "" {
		seq.FormatPattern = "{prefix}{year}{number:04d}"
	}
	if err := format.ValidatePattern(seq.FormatPattern); err != nil {
		return fmt.Errorf("validating format pattern: %w", err)
	}

	// Fetch existing for audit logging.
	existing, err := s.repo.GetByID(ctx, companyID, seq.ID)
	if err != nil {
		return fmt.Errorf("fetching existing sequence: %w", err)
	}

	// Prevent lowering next_number below already-used numbers.
	maxUsed, err := s.repo.MaxUsedNumber(ctx, companyID, seq.ID)
	if err != nil {
		return fmt.Errorf("checking max used number: %w", err)
	}
	if seq.NextNumber <= maxUsed {
		return fmt.Errorf("cannot set next number to %d, numbers up to %d have already been assigned: %w", seq.NextNumber, maxUsed, domain.ErrInvalidInput)
	}

	// Check uniqueness of prefix+year (within company) if they changed.
	duplicate, err := s.repo.GetByPrefixAndYear(ctx, companyID, seq.Prefix, seq.Year)
	if err == nil && duplicate != nil && duplicate.ID != seq.ID {
		return fmt.Errorf("sequence with prefix %q and year %d already exists: %w", seq.Prefix, seq.Year, domain.ErrDuplicateNumber)
	}

	if err := s.repo.Update(ctx, companyID, seq); err != nil {
		return fmt.Errorf("updating sequence: %w", err)
	}
	if s.audit != nil {
		s.audit.Log(ctx, "sequence", seq.ID, "update", existing, seq)
	}
	return nil
}

// Delete removes an invoice sequence by ID (soft delete) within the given company.
// Returns an error if invoices reference this sequence.
func (s *SequenceService) Delete(ctx context.Context, companyID, id int64) error {
	if id == 0 {
		return fmt.Errorf("sequence ID is required: %w", domain.ErrInvalidInput)
	}

	count, err := s.repo.CountInvoicesBySequenceID(ctx, companyID, id)
	if err != nil {
		return fmt.Errorf("checking invoice references: %w", err)
	}
	if count > 0 {
		return fmt.Errorf("cannot delete sequence: %d invoices reference it: %w", count, domain.ErrInvalidInput)
	}

	if err := s.repo.Delete(ctx, companyID, id); err != nil {
		return fmt.Errorf("deleting sequence: %w", err)
	}
	if s.audit != nil {
		s.audit.Log(ctx, "sequence", id, "delete", nil, nil)
	}
	return nil
}

// GetByID retrieves a sequence by its ID within the given company.
func (s *SequenceService) GetByID(ctx context.Context, companyID, id int64) (*domain.InvoiceSequence, error) {
	if id == 0 {
		return nil, fmt.Errorf("sequence ID is required: %w", domain.ErrInvalidInput)
	}
	seq, err := s.repo.GetByID(ctx, companyID, id)
	if err != nil {
		return nil, fmt.Errorf("fetching sequence: %w", err)
	}
	return seq, nil
}

// List retrieves all invoice sequences for the given company.
func (s *SequenceService) List(ctx context.Context, companyID int64) ([]domain.InvoiceSequence, error) {
	seqs, err := s.repo.List(ctx, companyID)
	if err != nil {
		return nil, fmt.Errorf("listing sequences: %w", err)
	}
	return seqs, nil
}

// GetOrCreateForYear retrieves an existing sequence for the given prefix and year
// within the given company, or creates a new one if it doesn't exist.
func (s *SequenceService) GetOrCreateForYear(ctx context.Context, companyID int64, prefix string, year int) (*domain.InvoiceSequence, error) {
	if prefix == "" {
		return nil, fmt.Errorf("prefix is required: %w", domain.ErrInvalidInput)
	}
	if year == 0 {
		return nil, fmt.Errorf("year is required: %w", domain.ErrInvalidInput)
	}

	seq, err := s.repo.GetByPrefixAndYear(ctx, companyID, prefix, year)
	if err == nil {
		return seq, nil
	}

	// Not found -- create a new sequence for this year.
	if !errors.Is(err, domain.ErrNotFound) {
		return nil, fmt.Errorf("looking up sequence: %w", err)
	}

	newSeq := &domain.InvoiceSequence{
		Prefix:        prefix,
		NextNumber:    1,
		Year:          year,
		FormatPattern: "{prefix}{year}{number:04d}",
	}
	if err := s.repo.Create(ctx, companyID, newSeq); err != nil {
		// Race condition: another goroutine may have created the sequence concurrently.
		// Retry the lookup; if found, use that sequence.
		retrySeq, retryErr := s.repo.GetByPrefixAndYear(ctx, companyID, prefix, year)
		if retryErr == nil {
			return retrySeq, nil
		}
		return nil, fmt.Errorf("creating sequence for year %d: %w", year, err)
	}
	return newSeq, nil
}

// FormatPreview returns the formatted invoice number the sequence would
// produce for its current next_number. Backed by internal/format.Render so
// preview, persistence, and GetNextNumber stay in lockstep. The empty-pattern
// fallback mirrors Create/Update — every layer normalises empty to the legacy
// default so renders never see an empty pattern.
func FormatPreview(seq *domain.InvoiceSequence) string {
	pattern := seq.FormatPattern
	if pattern == "" {
		pattern = "{prefix}{year}{number:04d}"
	}
	return format.Render(pattern, seq.Prefix, seq.Year, seq.NextNumber)
}

// SequenceCompanyChecker reports whether a company has any non-deleted
// invoice sequences. It satisfies the EntityChecker interface so
// CompanyService.Delete can refuse to soft-delete a company that still
// owns sequences.
type SequenceCompanyChecker struct {
	repo repository.InvoiceSequenceRepo
}

// NewSequenceCompanyChecker creates a new SequenceCompanyChecker.
func NewSequenceCompanyChecker(repo repository.InvoiceSequenceRepo) *SequenceCompanyChecker {
	return &SequenceCompanyChecker{repo: repo}
}

// CountNonDeletedForCompany returns the number of non-deleted invoice
// sequences belonging to the given company.
func (c *SequenceCompanyChecker) CountNonDeletedForCompany(ctx context.Context, companyID int64) (int, error) {
	list, err := c.repo.List(ctx, companyID)
	if err != nil {
		return 0, fmt.Errorf("counting sequences for company %d: %w", companyID, err)
	}
	return len(list), nil
}

File: rust/crates/zfaktury-core/src/service/sequence_svc.rs
use std::sync::Arc;

use zfaktury_domain::{DomainError, InvoiceSequence};

use super::audit_svc::AuditService;
use crate::repository::traits::InvoiceSequenceRepo;

/// Service for invoice numbering sequence management.
pub struct SequenceService {
    repo: Arc<dyn InvoiceSequenceRepo + Send + Sync>,
    audit: Option<Arc<AuditService>>,
}

impl SequenceService {
    pub fn new(
        repo: Arc<dyn InvoiceSequenceRepo + Send + Sync>,
        audit: Option<Arc<AuditService>>,
    ) -> Self {
        Self { repo, audit }
    }

    /// Creates a new invoice sequence. Validates prefix+year uniqueness.
    pub fn create(&self, seq: &mut InvoiceSequence) -> Result<(), DomainError> {
        if seq.prefix.is_empty() {
            return Err(DomainError::InvalidInput);
        }
        if seq.year == 0 {
            return Err(DomainError::InvalidInput);
        }
        if seq.next_number <= 0 {
            seq.next_number = 1;
        }
        if seq.format_pattern.is_empty() {
            seq.format_pattern = "{prefix}{year}{number:04d}".to_string();
        }

        // Check uniqueness of prefix+year.
        match self.repo.get_by_prefix_and_year(&seq.prefix, seq.year) {
            Ok(_) => return Err(DomainError::DuplicateNumber),
            Err(DomainError::NotFound) => {} // expected
            Err(e) => return Err(e),
        }

        self.repo.create(seq)?;
        if let Some(ref audit) = self.audit {
            audit.log("sequence", seq.id, "create", None, None);
        }
        Ok(())
    }

    /// Updates an existing sequence. Prevents lowering next_number below used numbers.
    pub fn update(&self, seq: &mut InvoiceSequence) -> Result<(), DomainError> {
        if seq.id == 0 || seq.prefix.is_empty() || seq.year == 0 || seq.next_number <= 0 {
            return Err(DomainError::InvalidInput);
        }

        self.repo.get_by_id(seq.id)?; // verify exists

        let max_used = self.repo.max_used_number(seq.id)?;
        if seq.next_number as i64 <= max_used {
            return Err(DomainError::InvalidInput);
        }

        // Check uniqueness of prefix+year (excluding self).
        if let Ok(dup) = self.repo.get_by_prefix_and_year(&seq.prefix, seq.year)
            && dup.id != seq.id
        {
            return Err(DomainError::DuplicateNumber);
        }

        self.repo.update(seq)?;
        if let Some(ref audit) = self.audit {
            audit.log("sequence", seq.id, "update", None, None);
        }
        Ok(())
    }

    /// Deletes a sequence if no invoices reference it.
    pub fn delete(&self, id: i64) -> Result<(), DomainError> {
        if id == 0 {
            return Err(DomainError::InvalidInput);
        }

        let count = self.repo.count_invoices_by_sequence_id(id)?;
        if count > 0 {
            return Err(DomainError::InvalidInput);
        }

        self.repo.delete(id)?;
        if let Some(ref audit) = self.audit {
            audit.log("sequence", id, "delete", None, None);
        }
        Ok(())
    }

    pub fn get_by_id(&self, id: i64) -> Result<InvoiceSequence, DomainError> {
        if id == 0 {
            return Err(DomainError::InvalidInput);
        }
        self.repo.get_by_id(id)
    }

    pub fn list(&self) -> Result<Vec<InvoiceSequence>, DomainError> {
        self.repo.list()
    }

    /// Retrieves or creates a sequence for the given prefix and year.
    pub fn get_or_create_for_year(
        &self,
        prefix: &str,
        year: i32,
    ) -> Result<InvoiceSequence, DomainError> {
        if prefix.is_empty() || year == 0 {
            return Err(DomainError::InvalidInput);
        }

        match self.repo.get_by_prefix_and_year(prefix, year) {
            Ok(seq) => return Ok(seq),
            Err(DomainError::NotFound) => {}
            Err(e) => return Err(e),
        }

        let mut new_seq = InvoiceSequence {
            id: 0,
            prefix: prefix.to_string(),
            next_number: 1,
            year,
            format_pattern: "{prefix}{year}{number:04d}".to_string(),
        };

        match self.repo.create(&mut new_seq) {
            Ok(()) => Ok(new_seq),
            Err(_) => {
                // Race condition: retry lookup.
                self.repo.get_by_prefix_and_year(prefix, year)
            }
        }
    }

    /// Returns a preview of the next formatted invoice number.
    pub fn format_preview(seq: &InvoiceSequence) -> String {
        format!("{}{}{:04}", seq.prefix, seq.year, seq.next_number)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::Mutex;

    struct MockSeqRepo {
        seqs: Mutex<Vec<InvoiceSequence>>,
        next_id: Mutex<i64>,
    }

    impl MockSeqRepo {
        fn new() -> Self {
            Self {
                seqs: Mutex::new(Vec::new()),
                next_id: Mutex::new(1),
            }
        }
    }

    impl InvoiceSequenceRepo for MockSeqRepo {
        fn create(&self, seq: &mut InvoiceSequence) -> Result<(), DomainError> {
            let mut id = self.next_id.lock().unwrap();
            seq.id = *id;
            *id += 1;
            self.seqs.lock().unwrap().push(seq.clone());
            Ok(())
        }

        fn update(&self, seq: &mut InvoiceSequence) -> Result<(), DomainError> {
            let mut seqs = self.seqs.lock().unwrap();
            if let Some(s) = seqs.iter_mut().find(|s| s.id == seq.id) {
                *s = seq.clone();
                Ok(())
            } else {
                Err(DomainError::NotFound)
            }
        }

        fn delete(&self, id: i64) -> Result<(), DomainError> {
            let mut seqs = self.seqs.lock().unwrap();
            seqs.retain(|s| s.id != id);
            Ok(())
        }

        fn get_by_id(&self, id: i64) -> Result<InvoiceSequence, DomainError> {
            self.seqs
                .lock()
                .unwrap()
                .iter()
                .find(|s| s.id == id)
                .cloned()
                .ok_or(DomainError::NotFound)
        }

        fn list(&self) -> Result<Vec<InvoiceSequence>, DomainError> {
            Ok(self.seqs.lock().unwrap().clone())
        }

        fn get_by_prefix_and_year(
            &self,
            prefix: &str,
            year: i32,
        ) -> Result<InvoiceSequence, DomainError> {
            self.seqs
                .lock()
                .unwrap()
                .iter()
                .find(|s| s.prefix == prefix && s.year == year)
                .cloned()
                .ok_or(DomainError::NotFound)
        }

        fn count_invoices_by_sequence_id(&self, _sequence_id: i64) -> Result<i64, DomainError> {
            Ok(0)
        }

        fn max_used_number(&self, _sequence_id: i64) -> Result<i64, DomainError> {
            Ok(0)
        }
    }

    #[test]
    fn test_create_sequence() {
        let repo = Arc::new(MockSeqRepo::new());
        let svc = SequenceService::new(repo, None);

        let mut seq = InvoiceSequence {
            id: 0,
            prefix: "FV".to_string(),
            next_number: 1,
            year: 2024,
            format_pattern: String::new(),
        };
        svc.create(&mut seq).unwrap();
        assert!(seq.id > 0);
        assert!(!seq.format_pattern.is_empty());
    }

    #[test]
    fn test_duplicate_prefix_year_rejected() {
        let repo = Arc::new(MockSeqRepo::new());
        let svc = SequenceService::new(repo, None);

        let mut s1 = InvoiceSequence {
            id: 0,
            prefix: "FV".to_string(),
            next_number: 1,
            year: 2024,
            format_pattern: String::new(),
        };
        svc.create(&mut s1).unwrap();

        let mut s2 = InvoiceSequence {
            id: 0,
            prefix: "FV".to_string(),
            next_number: 1,
            year: 2024,
            format_pattern: String::new(),
        };
        assert!(matches!(
            svc.create(&mut s2),
            Err(DomainError::DuplicateNumber)
        ));
    }

    #[test]
    fn test_get_or_create() {
        let repo = Arc::new(MockSeqRepo::new());
        let svc = SequenceService::new(repo, None);

        let seq1 = svc.get_or_create_for_year("FV", 2024).unwrap();
        let seq2 = svc.get_or_create_for_year("FV", 2024).unwrap();
        assert_eq!(seq1.id, seq2.id);
    }

    #[test]
    fn test_format_preview() {
        let seq = InvoiceSequence {
            id: 1,
            prefix: "FV".to_string(),
            next_number: 5,
            year: 2024,
            format_pattern: String::new(),
        };
        assert_eq!(SequenceService::format_preview(&seq), "FV20240005");
    }
}
